src/asyncresults.jl

```diff
@@ -0,0 +1,256 @@
+"""
+function cancel(async_result::AsyncResult)
+    async_result.should_cancel = true
+    return
+end
+
+function _cancel(jl_conn::Connection)
+    cancel_ptr = libpq_c.PQgetCancel(jl_conn.conn)
+    try
+        # https://www.postgresql.org/docs/10/libpq-cancel.html#LIBPQ-PQCANCEL
+        errbuf_size = 256
+        errbuf = fill(0x0, errbuf_size)
+        success = libpq_c.PQcancel(cancel_ptr, pointer(errbuf), errbuf_size) == 1
+        if !success
+            warn(LOGGER, "Failed cancelling query: $(String(errbuf))")
+        else
+            debug(LOGGER, "Cancelled query for connection $(jl_conn.conn)")
+        end
+    finally
+        libpq_c.PQfreeCancel(cancel_ptr)
+    end
+end
+
+iserror(async_result::AsyncResult) = Base.istaskfailed(async_result.result_task)
+Base.isready(async_result::AsyncResult) = istaskdone(async_result.result_task)
+Base.wait(async_result::AsyncResult) = wait(async_result.result_task)
+Base.fetch(async_result::AsyncResult) = fetch(async_result.result_task)
+trywait(async_result::AsyncResult) = (try wait(async_result) catch end; nothing)
+trywait(::Nothing) = nothing
+Base.close(async_result::AsyncResult) = cancel(async_result)
+
+"""
+    async_execute(
+        jl_conn::Connection,
+        query::AbstractString,
+        [parameters::Union{AbstractVector, Tuple},]
+        kwargs...
+    ) -> AsyncResult
+
+Run a query on the PostgreSQL database and return an [`AsyncResult`](@ref).
+
+The `AsyncResult` contains a `Task` which processes a query asynchronously.
+Calling `fetch` on the `AsyncResult` will return a [`Result`](@ref).
+
+All keyword arguments are the same as [`execute`](@ref) and are passed to the created
+`Result`.
+
+`async_execute` does not yet support [`Statement`](@ref)s.
+
+`async_execute` optionally takes a `parameters` vector which passes query parameters as
+strings to PostgreSQL.
+Queries without parameters can contain multiple SQL statements, and the result of the final
+statement is returned.
+Any errors which occur during executed statements will be bundled together in a
+`CompositeException` and thrown.
+
+As is normal for `Task`s, any exceptions will be thrown when calling `wait` or `fetch`.
+"""
+function async_execute end
+
+function async_execute(jl_conn::Connection, query::AbstractString; kwargs...)
+    async_result = _async_execute(jl_conn; kwargs...) do jl_conn
+        _async_submit(jl_conn.conn, query)
+    end
+
+    return async_result
+end
+
+function async_execute(
+    jl_conn::Connection,
+    query::AbstractString,
+    parameters::Union{AbstractVector, Tuple};
+    kwargs...
+)
+    string_params = string_parameters(parameters)
+    pointer_params = parameter_pointers(string_params)
+
+    async_result = _async_execute(jl_conn; kwargs...) do jl_conn
+        _async_submit(jl_conn.conn, query, pointer_params)
+    end
+
+    return async_result
+end
+
+function _async_execute(
+    submission_fn::Base.Callable, jl_conn::Connection; throw_error::Bool=true, kwargs...
```

Comment:
How about dropping the constraint
```suggestion
    submission_fn, jl_conn::Connection; throw_error::Bool=true, kwargs...
```

This is still nonambigiouis, and now we allow any functor,
and avoid using the `Base.Callable` which is not part of the public API.

Though on the other hand, we only use this internally, with an anon function,
so we could also do
```suggestion
    submission_fn::Function, jl_conn::Connection; throw_error::Bool=true, kwargs...
```

if the intent was to have a self documenting signature